proposals/0000-package-internals.rst

```diff
@@ -0,0 +1,148 @@
+   only on ``base`` (which re-exports much, but not all, of ``ghc-prim``). This
+   proposal is intended to protect users from unknowingly depending on ``ghc-prim``
+   modules, which might change frequently.
+
+Proposed Change Specification
+-----------------------------
+
+1. Add a new command-line flag ``-expose-package-internals <pkg>``, where ``<pkg>`` is understood
+   in the same way as the argument to the existing ``-package`` flag.
+
+#. It is an error for the ``<pkg>`` named in ``-expose-package-internals`` not to be exposed.
+
+#. Imports of modules from a package named in ``-expose-package-internals`` skip the exposed-module check.
+
+#. The error GHC produces when a non-exposed module is imported warns the user about
+   exposing internal modules, with something like ::
+
+    Could not load module 'Data.OldList';
+    it is a hidden module in the package 'base-4.15.1.0'.
+    If you wish to be able to access an internal module anyway, you may
+    specify '-expose-package-internals base-4.15.1.0'; however, be warned
+    that the API from internal modules may change between minor version releases.
+
+#. The `Haskell Package Versioning Policy <https://pvp.haskell.org/>`_ specifies that changes in
+   internal modules do *not* require a major version change.
+
+#. Haddock output does not change: only exposed modules are included in the documentation.
+   Users wishing to depend on internal modules are expected to consult the source code.
+
+#. For the next three releases of GHC, any import of a module exported from ``ghc-prim``
+   will emit a warning (controlled by ``-Wghc-prim-exports`` and part of ``-Wcompat``),
+   unless ``-expose-package-internals ghc-prim`` is specified. The warning will inform
+   users that all modules in ``ghc-prim`` will become internal after three releases.
+   When that happens, the warning itself will also be removed.
+
+
+Examples
+--------
+
+*Example.hs*\ ::
+
+  module Example (foldl) where
+  import Data.OldList   -- internal module in `base`
+
+::
+
+  > ghc -c Example.hs
+  Example.hs:2:1: error:
+    Could not load module 'Data.OldList';
+    it is a hidden module in the package 'base-4.15.1.0'.
+    If you wish to be able to access an internal module anyway, you may
+    specify '-expose-package-internals base-4.15.1.0'; however, be warned
+    that the API from internal modules may change between minor version releases.
+
+  > ghc -c Example.hs -expose-package-internals base
+  > ls
+  Example.hs Example.o Example.hi
+
+Effect and Interactions
+-----------------------
+
+1. All GHC users can now access internal modules from other packages, improving
+   the situations outlined in the Motivation_.
+
+#. The exports from ``ghc-prim`` are now protected; users can discover that
+   using names from ``ghc-prim`` is potentially troublesome.
+
+#. There may be a desire to integrate support for ``-expose-package-internals`` in ``.cabal``
```

Comment:
Probably not even needed, the `ghc-options` stanza would just work out of the box here I assume?

Reply:
Not necessarily, I think. Maybe there are cases where the same package is included twice but with different unit-ids (either different versions of the same package or different dependencies), and then you'd want to supply the exactly-correct unit-id. Maybe this case is impossible? But having this info directly in the cabal file would also allow more inspection of it, in case a library does not want to depend on other libraries that depend on internals (because then the original library's dependencies might break unexpectedly).